modules/index/src/main/java/org/apache/ignite/internal/index/IndexCollector.java

```diff
@@ -0,0 +1,296 @@
+                return tableIndexes;
+            }
+
+            return unmodifiableList(merge(tableIndexes, droppedAvailableTableIndexes));
+        });
+    }
+
+    private void addListeners() {
+        catalogService.listen(CatalogEvent.INDEX_DROP, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onDropIndex((DropIndexEventParameters) parameters).thenApply(unused -> false);
+        });
+
+        catalogService.listen(CatalogEvent.TABLE_DROP, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onTableDrop((DropTableEventParameters) parameters).thenApply(unused -> false);
+        });
+    }
+
+    private CompletableFuture<?> onDropIndex(DropIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            int previousCatalogVersion = parameters.catalogVersion() - 1;
+
+            CatalogIndexDescriptor droppedIndexDescriptor = catalogService.index(parameters.indexId(), previousCatalogVersion);
+
+            assert droppedIndexDescriptor != null : "indexId=" + parameters.indexId() + ", catalogVersion=" + previousCatalogVersion;
+
+            if (!droppedIndexDescriptor.available()) {
+                return completedFuture(null);
+            }
+
+            addDroppedAvailableIndex(droppedIndexDescriptor, parameters.catalogVersion());
+
+            return completedFuture(null);
+        });
+    }
+
+    private CompletableFuture<?> onTableDrop(DropTableEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            droppedAvailableTableIndexes.entrySet().removeIf(entry -> parameters.tableId() == entry.getKey().tableId);
+
+            return completedFuture(null);
+        });
+    }
+
+    /**
+     * Returns a list of dropped available indexes (sorted by {@link #INDEX_COMPARATOR}) for the catalog version of interest from
+     * {@link #droppedAvailableTableIndexes}. If there is no list for the requested catalog version, the closest previous catalog version
+     * will be returned.
+     */
+    private List<CatalogIndexDescriptor> getDroppedAvailableIndexes(int catalogVersion, int tableId) {
+        var key = new TableIdCatalogVersion(tableId, catalogVersion);
+
+        Entry<TableIdCatalogVersion, List<CatalogIndexDescriptor>> entry = droppedAvailableTableIndexes.floorEntry(key);
+
+        return entry != null && tableId == entry.getKey().tableId ? entry.getValue() : List.of();
+    }
+
+    /**
+     * Adds the dropped available index to {@link #droppedAvailableTableIndexes}.
+     *
+     * <p>If the list is missing for the catalog version from the arguments, then we create it by merging the indexes from the previous
+     * catalog version and the new index. Otherwise, we simply add to the existing list. List are sorted by {@link #INDEX_COMPARATOR}</p>
```

Comment:
```suggestion
     * catalog version and the new index. Otherwise, we simply add to the existing list. Lists are sorted by {@link #INDEX_COMPARATOR}</p>
```

Reply:
fix it